<!-- -*- mode: XML -*- -->

<launch>
   <arg name="simu" default="false"/>
   <arg name="world_name" default="/usr/share/gazebo-2.2/worlds/willowgarage.world"/>
   <!-- <arg name="world_name" default="$(find racecar)/worlds/maze.sdf"/> -->
   <arg name="waypoint_check_distance" default="0.5"/>
   <arg name="run_camera" default="true"/>

  <!-- Robot Description for Rviz visualization -->
    <param name="robot_description"
           command="$(find xacro)/xacro.py $(find gt_car_vehicle_description)/urdf/gt_car.urdf.xacro"/>
<node pkg="racecar" type="joint_publisher" name="joint_publisher" output="screen">
    <param name="frequency" value="100" />
  </node>
    <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_pub" output="screen"/>


  <!-- Group to launch to simulate the system -->
  <group if="$(arg simu)">
    <node pkg="tf" type="static_transform_publisher" name="base_link_to_laser_box" 
    args="0.0 0.0 0.0 0.0 0.0 0.0 /base_link /gt_car_vehicle/laser_box 40" />
    <include file="$(find gt_car_vehicle_description)/launch/gazebo.launch">
	<arg name="world_name" value="$(arg world_name)"/>
    </include>
    <node pkg="racecar" type="servo_command_to_simu" name="servo_command_to_simu" output="screen">
      <param name="frequency" value="100" />
      <param name="alpha" value="0.3" />
      <param name="linear_decay" value="0.5" />
      <param name="delta_max" value="0.7" />
      <param name="v_max" value="1.0" />
    </node>
    <node pkg="rviz" type="rviz" name="rviz" args="-d $(find racecar)/rviz/laser_scan_matcher.rviz"/>
  </group>

  <!-- Group to launch the real system -->
  <group unless="$(arg simu)">
    <include file="$(find racecar)/launch/includes/joy_estop.launch.xml" />
    <include file="$(find racecar)/launch/includes/steering.launch.xml" />
    <include file="$(find racecar)/launch/includes/throttle.launch.xml" />
    <include file="$(find racecar)/launch/includes/sensors.launch.xml">
      <arg name="run_camera" value="$(arg run_camera)" />
    </include> 
  </group>
    
  <!-- Nodes that are launched for both the simulation and the real system -->

  <node pkg="rosbag" type="record" name="bag_recorder"
	args="record -O /media/DATA/bags/last_run.bag -a" /> 
	
  <node pkg="racecar" type="laser_to_estop" name="laser_to_estop" output="screen">
    <param name="filter_alpha" value="0.95" />
    <param name="max_min_dist" value="0.2" />
    <param name="max_impact_time" value="0.9" />
  </node>
  
  <node pkg="racecar" type="goal_marker_node" name="goal_marker_node" output="screen"/>
  
  <node pkg="racecar" type="waypoint_tracker" name="waypoint_tracker" output="screen">
    <param name="filter_alpha" value="0.8" />
    <param name="frequency" value="100.0" />
    <param name="speed_max" value="3.0" />
    <param name="throttle_max" value="1.0" />
    <param name="throttle_min" value="0.85" />
    <param name="throttle_dist_gain" value="1.0" />
    <param name="throttle_speed_gain" value="1.5" />
    <param name="steering_gain" value="1.0" />
    <param name="uturn_throttle" value="1.0" />
    <param name="waypoint_check_distance" value="$(arg waypoint_check_distance)" />
  </node>

  <node pkg="racecar" type="planner" name="planner" output="screen">
    <param name="tree_depth" value="10" />
    <param name="nb_obstacle_check" value="20" />
    <param name="epsilon" value="0.5" />
    <param name="unknown_space_probability" value="0.4" />
    <param name="inflation_radius" value="0.4" />
    <param name="waypoint_check_distance" value="$(arg waypoint_check_distance)" />
    <param name="waypoint_max_distance" value="0.2" />
    <param name="keep_moving" value="false" />
  </node>

  <include file="$(find racecar)/launch/mapping.launch" />

</launch>
